File: tray/service-manager.js
'use strict';

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

const HEARTBEAT_DIR = path.join(os.homedir(), '.config', 'claudia', 'heartbeats');

// Map launchd labels to heartbeat service names
const HEARTBEAT_NAMES = {
  'ai.claudia.gmail-monitor': 'gmail-monitor',
  'ai.claudia.slack-events': 'slack-events',
  'ai.claudia.meeting-alerts': 'meeting-alerts',
  'ai.claudia.followup-checker': 'followup-checker',
  'ai.openclaw.gateway': null,  // Gateway doesn't write heartbeats (yet)
};

const SERVICES = [
  { label: 'Gmail Monitor',      launchdLabel: 'ai.claudia.gmail-monitor' },
  { label: 'Slack Events',       launchdLabel: 'ai.claudia.slack-events' },
  { label: 'Meeting Alerts',     launchdLabel: 'ai.claudia.meeting-alerts' },
  { label: 'Follow-up Checker',  launchdLabel: 'ai.claudia.followup-checker' },
  { label: 'Gateway',            launchdLabel: 'ai.openclaw.gateway' },
];

function parseLaunchctlList(output) {
  const map = {};
  for (const line of output.split('\n')) {
    const parts = line.trim().split(/\t/);
    if (parts.length < 3) continue;
    const [pidStr, exitStr, label] = parts;
    if (label === 'Label') continue;
    map[label] = {
      pid: pidStr === '-' ? null : parseInt(pidStr, 10),
      exitCode: parseInt(exitStr, 10)
    };
  }
  return map;
}

function statusFromEntry(entry) {
  if (!entry) return 'unloaded';
  if (entry.pid) return 'running';
  return entry.exitCode === 0 ? 'stopped' : 'error';
}

let cachedUID = null;
function getUID() {
  if (!cachedUID) cachedUID = execSync('id -u', { encoding: 'utf8' }).trim();
  return cachedUID;
}

function readHeartbeat(launchdLabel) {
  const name = HEARTBEAT_NAMES[launchdLabel];
  if (!name) return null;
  try {
    const raw = fs.readFileSync(path.join(HEARTBEAT_DIR, `${name}.json`), 'utf8');
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

function evaluateHeartbeat(hb) {
  if (!hb) return { health: 'unknown', detail: null, errorCount: 0 };
  if (hb.status === 'startup-failed') {
    return { health: 'startup-failed', detail: hb.errors ? hb.errors.lastError : 'Unknown error', errorCount: 0 };
  }
  if (hb.status === 'error' || hb.status === 'degraded') {
    return { health: hb.status, detail: hb.errors ? hb.errors.lastError : null, errorCount: hb.errors ? hb.errors.countSinceStart : 0 };
  }
  if (hb.status === 'shutting-down') {
    return { health: 'shutting-down', detail: null, errorCount: 0 };
  }
  const age = Date.now() - hb.lastCheck;
  if (age > hb.checkInterval * 3) {
    return { health: 'unresponsive', detail: `No heartbeat for ${Math.round(age / 60000)}m`, errorCount: 0 };
  }
  return { health: 'healthy', detail: null, errorCount: hb.errors ? hb.errors.countSinceStart : 0 };
}

function getStatuses() {
  let output;
  try {
    output = execSync('launchctl list', { encoding: 'utf8', timeout: 5000 });
  } catch (e) {
    return SERVICES.map(s => ({ ...s, status: 'unknown', pid: null, exitCode: null, heartbeat: null, heartbeatHealth: null }));
  }
  const parsed = parseLaunchctlList(output);
  return SERVICES.map(svc => {
    const entry = parsed[svc.launchdLabel];
    const hb = readHeartbeat(svc.launchdLabel);
    const hbHealth = evaluateHeartbeat(hb);
    return {
      ...svc,
      status: statusFromEntry(entry),
      pid: entry ? entry.pid : null,
      exitCode: entry ? entry.exitCode : null,
      heartbeat: hb,
      heartbeatHealth: hbHealth
    };
  });
}

function startService(launchdLabel) {
  execSync(`launchctl kickstart gui/${getUID()}/${launchdLabel}`, { timeout: 5000 });
}

function stopService(launchdLabel) {
  execSync(`launchctl kill SIGTERM gui/${getUID()}/${launchdLabel}`, { timeout: 5000 });
}

function restartService(launchdLabel) {
  execSync(`launchctl kickstart -k gui/${getUID()}/${launchdLabel}`, { timeout: 5000 });
}

module.exports = { SERVICES, parseLaunchctlList, statusFromEntry, getStatuses, startService, stopService, restartService, readHeartbeat, evaluateHeartbeat };

File: reticle/Sources/Reticle/Services/ServiceManager.swift
import Foundation

// MARK: - Types

enum ServiceStatus: String, Equatable, Codable {
    case running, stopped, error, unloaded, unknown
}

struct LaunchctlEntry: Equatable {
    let pid: Int?
    let exitCode: Int
}

struct HeartbeatErrors: Codable {
    let lastError: String?
    let lastErrorAt: Double?
    let countSinceStart: Int
}

struct HeartbeatMetrics: Codable {
    let recording: Bool?
    let meetingId: String?
    let duration: Double?
    let captureMode: String?
    let permissionStatus: String?
    let itemCount: Int?
    let patternCount: Int?
    let degradedReason: String?
}

struct HeartbeatData: Codable {
    let service: String?
    let pid: Int?
    let startedAt: Double?
    let lastCheck: Double?
    let checkInterval: Double?
    let status: String?
    let errors: HeartbeatErrors?
    let metrics: HeartbeatMetrics?
}

struct HeartbeatHealth: Equatable {
    let health: String
    let detail: String?
    let errorCount: Int
    static let unknown = HeartbeatHealth(health: "unknown", detail: nil, errorCount: 0)
}

struct ServiceDefinition {
    let label: String
    let launchdLabel: String
    let heartbeatName: String?
    let scheduled: Bool

    static let all: [ServiceDefinition] = [
        ServiceDefinition(label: "Gmail Monitor",     launchdLabel: "ai.reticle.gmail-monitor",      heartbeatName: "gmail-monitor",      scheduled: false),
        ServiceDefinition(label: "Slack Events",      launchdLabel: "ai.reticle.slack-events",       heartbeatName: "slack-events",       scheduled: false),
        ServiceDefinition(label: "Meeting Alerts",    launchdLabel: "ai.reticle.meeting-alerts",     heartbeatName: "meeting-alerts",     scheduled: false),
        ServiceDefinition(label: "Follow-up Checker", launchdLabel: "ai.reticle.followup-checker",   heartbeatName: "followup-checker",   scheduled: false),
        ServiceDefinition(label: "Meeting Recorder",  launchdLabel: "ai.reticle.meeting-recorder",   heartbeatName: "meeting-recorder",   scheduled: false),
        ServiceDefinition(label: "Gateway",           launchdLabel: "ai.reticle.gateway",            heartbeatName: "gateway",            scheduled: false),
        ServiceDefinition(label: "Daily Digest",      launchdLabel: "ai.reticle.digest-daily",       heartbeatName: "digest-daily",       scheduled: true),
        ServiceDefinition(label: "Weekly Digest",     launchdLabel: "ai.reticle.digest-weekly",      heartbeatName: "digest-weekly",      scheduled: true),
    ]
}

struct ServiceState {
    let definition: ServiceDefinition
    let status: ServiceStatus
    let pid: Int?
    let exitCode: Int?
    let heartbeat: HeartbeatData?
    let heartbeatHealth: HeartbeatHealth
}

// MARK: - ServiceManager

class ServiceManager {

    private static let heartbeatDir: String = {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        return "\(home)/.reticle/heartbeats"
    }()

    private static var cachedUID: String?

    // MARK: - Pure functions

    static func parseLaunchctlList(_ output: String) -> [String: LaunchctlEntry] {
        var map: [String: LaunchctlEntry] = [:]
        for line in output.split(separator: "\n", omittingEmptySubsequences: true) {
            let parts = line.split(separator: "\t")
            guard parts.count >= 3 else { continue }
            let label = String(parts[2])
            if label == "Label" { continue }
            let pid = parts[0] == "-" ? nil : Int(parts[0])
            let exitCode = Int(parts[1]) ?? 0
            map[label] = LaunchctlEntry(pid: pid, exitCode: exitCode)
        }
        return map
    }

    static func statusFromEntry(_ entry: LaunchctlEntry?) -> ServiceStatus {
        guard let entry = entry else { return .unloaded }
        if entry.pid != nil { return .running }
        return entry.exitCode == 0 ? .stopped : .error
    }

    static func evaluateHeartbeat(_ hb: HeartbeatData?, now: Double? = nil) -> HeartbeatHealth {
        guard let hb = hb else { return .unknown }

        if hb.status == "startup-failed" {
            let detail = hb.errors?.lastError ?? "Unknown error"
            return HeartbeatHealth(health: "startup-failed", detail: detail, errorCount: 0)
        }
        if hb.status == "error" || hb.status == "degraded" {
            return HeartbeatHealth(
                health: hb.status!,
                detail: hb.errors?.lastError,
                errorCount: hb.errors?.countSinceStart ?? 0
            )
        }
        if hb.status == "shutting-down" {
            return HeartbeatHealth(health: "shutting-down", detail: nil, errorCount: 0)
        }

        // If checkInterval is present, check staleness
        if let lastCheck = hb.lastCheck, let interval = hb.checkInterval {
            let currentTime = now ?? (Date().timeIntervalSince1970 * 1000)
            let ageMs = currentTime - lastCheck
            if ageMs > interval * 3 {
                let ageMin = Int(ageMs / 60000)
                return HeartbeatHealth(health: "unresponsive", detail: "No heartbeat for \(ageMin)m", errorCount: 0)
            }
        }

        // Scheduled services may not write checkInterval — if lastCheck exists, treat as healthy
        if hb.lastCheck != nil {
            return HeartbeatHealth(
                health: "healthy",
                detail: nil,
                errorCount: hb.errors?.countSinceStart ?? 0
            )
        }

        return .unknown
    }

    // MARK: - I/O operations

    static func getUID() -> String {
        if let cached = cachedUID { return cached }
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/id")
        process.arguments = ["-u"]
        let pipe = Pipe()
        process.standardOutput = pipe
        try? process.run()
        process.waitUntilExit()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        let uid = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "501"
        cachedUID = uid
        return uid
    }

    static func readHeartbeat(launchdLabel: String) -> HeartbeatData? {
        // Find heartbeat name from service definitions
        guard let def = ServiceDefinition.all.first(where: { $0.launchdLabel == launchdLabel }),
              let name = def.heartbeatName else {
            return nil
        }
        let filePath = "\(heartbeatDir)/\(name).json"
        guard let data = FileManager.default.contents(atPath: filePath) else { return nil }
        return try? JSONDecoder().decode(HeartbeatData.self, from: data)
    }

    static func getStatuses() -> [ServiceState] {
        let output: String
        do {
            let process = Process()
            process.executableURL = URL(fileURLWithPath: "/bin/launchctl")
            process.arguments = ["list"]
            let pipe = Pipe()
            process.standardOutput = pipe
            try process.run()
            process.waitUntilExit()
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            output = String(data: data, encoding: .utf8) ?? ""
        } catch {
            return ServiceDefinition.all.map { def in
                ServiceState(
                    definition: def,
                    status: .unknown,
                    pid: nil,
                    exitCode: nil,
                    heartbeat: nil,
                    heartbeatHealth: .unknown
                )
            }
        }

        let parsed = parseLaunchctlList(output)

        return ServiceDefinition.all.map { def in
            let entry = parsed[def.launchdLabel]
            let hb = readHeartbeat(launchdLabel: def.launchdLabel)
            let hbHealth = evaluateHeartbeat(hb)
            return ServiceState(
                definition: def,
                status: statusFromEntry(entry),
                pid: entry?.pid,
                exitCode: entry?.exitCode,
                heartbeat: hb,
                heartbeatHealth: hbHealth
            )
        }
    }

    static func startService(_ launchdLabel: String) throws {
        let uid = getUID()
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/launchctl")
        process.arguments = ["kickstart", "gui/\(uid)/\(launchdLabel)"]
        try process.run()
        process.waitUntilExit()
        if process.terminationStatus != 0 {
            throw ServiceManagerError.launchctlFailed(label: launchdLabel, code: Int(process.terminationStatus))
        }
    }

    static func stopService(_ launchdLabel: String) throws {
        let uid = getUID()
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/launchctl")
        process.arguments = ["kill", "SIGTERM", "gui/\(uid)/\(launchdLabel)"]
        try process.run()
        process.waitUntilExit()
        if process.terminationStatus != 0 {
            throw ServiceManagerError.launchctlFailed(label: launchdLabel, code: Int(process.terminationStatus))
        }
    }

    static func restartService(_ launchdLabel: String) throws {
        let uid = getUID()
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/launchctl")
        process.arguments = ["kickstart", "-k", "gui/\(uid)/\(launchdLabel)"]
        try process.run()
        process.waitUntilExit()
        if process.terminationStatus != 0 {
            throw ServiceManagerError.launchctlFailed(label: launchdLabel, code: Int(process.terminationStatus))
        }
    }
}

enum ServiceManagerError: Error, LocalizedError {
    case launchctlFailed(label: String, code: Int)

    var errorDescription: String? {
        switch self {
        case .launchctlFailed(let label, let code):
            return "launchctl failed for \(label) with exit code \(code)"
        }
    }
}
